TM1py/Objects/GitProject.py

```diff
@@ -1,42 +1,116 @@
 # -*- coding: utf-8 -*-
 
 import json
-from typing import Optional, Iterable, Dict, List, Union
+from typing import Optional, Dict, List
+
 from TM1py.Objects.TM1Object import TM1Object
 
 
-def cleanNullTerms(d: dict):
+def clean_null_terms(d: Dict):
     clean = {}
     for k, v in d.items():
         if isinstance(v, dict):
-            nested = cleanNullTerms(v)
+            nested = clean_null_terms(v)
             if len(nested.keys()) > 0:
                 clean[k] = nested
         elif v is not None:
             clean[k] = v
     return clean
 
 
+class TM1ProjectTask:
+
+    def __init__(self, task_name: str, chore: str = None, process: str = None, parameters: List[Dict[str, str]] = None,
+                 dependencies: List[str] = None, precondition: str = None):
+        """
+        Defines an action that executes a Process or a Chore with certain parameters.
+
+        A Task MUST either have a Process or a Chore property.
+        The property specifies the reference of the Process or Chore to be executed.
+        The Process or Chore MUST be visible.
+
+        A Task MAY have a Parameters property.
+        The property specifies the parameters to be passed to the Process.
+        This property MUST NOT be specified if the task is to execute a Chore.
+
+        A Task MAY have a Dependencies property.
+        The property specifies an array of URIs of tasks or objects,
+        which will be executed or loaded, respectively, before executing the current task.
+        E.g.: ["Cubes('Cube_A')", "Dimensions('Dimension_C')"]
+
+        A Task MAY have a Precondition property.
+        The server only executes a Task when either the precondition is not specified, or it is evaluated to TRUE.
+
+        The server only executes a Task one time during a deployment.
+        """
+        if precondition:

```

Comment:
Thanks. I agree. We would need helper functions to remove from "tasks" and "ignore" as well.

For the preconditions, I'm not sure. I guess we could allow simple string values if they were passed.

How do you imagine people use the TM1ProjectDeployment class. Can you post a code sample?

Reply:
I found the deployment in the IBM docs. Yeah this is cool. It makes sense to implement it as a subclass of TM1Project


> Deployment
> Deployment is an OPTIONAL property. Each of its property defines a named deployment and its specific properties.
> 
> All the tm1project properties can be redefined for a deployment, except Version. Those properties overrides the tm1project properties for the specific deployment. Current deployment is set by action GitInit.

Reply:
I agree.

I will change the TM1ProjectTask to allow for preconditions.

Are you saying that preconditions can also be defined on a project level? I couldn't find anything about this in the docs.

Reply:
> I found the deployment in the IBM docs. Yeah this is cool. It makes sense to implement it as a subclass of TM1Project
> 
> > Deployment
> > Deployment is an OPTIONAL property. Each of its property defines a named deployment and its specific properties.
> > All the tm1project properties can be redefined for a deployment, except Version. Those properties overrides the tm1project properties for the specific deployment. Current deployment is set by action GitInit.

Yep, this is actually the only reason why I would mess with tm1project. I think this will be particularly useful for re-defining both ‘Settings' (Performance Monitor, TIDebugger, etc.) and ‘PrePull’,’PostPull’ (savedataall, kickoff users, backup the instance, etc)  per environment.